docs/design-documents/drivers/serial/serial.md

```diff
@@ -0,0 +1,317 @@
+                   void *tx_buffer,
+                   size_t tx_buffer_size);
+```
+
+ The rest of the class API is that of `UARTSerial` and is described [here](https://os.mbed.com/docs/mbed-os/v5.14/mbed-os-api-doxy/classmbed_1_1_u_a_r_t_serial.html).
+
+
+### Detailed design : UnbufferedSerial
+
+`UnbufferedSerial` is a new class that adds a `FileHandle` interface to `RawSerial`. `RawSerial` is the preferred implementation for unbuffered I/O but it lacks a `FileHandle` interface and so it cannot be used by the retarget code. Instead the retarget code uses a non-public `DirectSerial` class which adds to the complexity of classes and should be removed. `UnbufferedSerial` will not provide any of the printf methods that `RawSerial` has; applications should use _fdopen_ to get a _FILE*_ to use the printf function, see [Usage scenarios and examples](#usage-scenarios-and-examples).
+
+The below diagram shows the detailed inheritance hierarchy for `UnbufferedSerial`.
+
+![UnbufferedSerial class](./diagrams/unbuffered-serial.png)
+
+**API description**
+
+This is the proposed API for the `UnbufferedSerial` class.
+
+```C
+class UnbufferedSerial: public SerialBase, public FileHandle, private NonCopyable<UnbufferedSerial> {
+
+public:
+    /** Create a UnbufferedSerial port, connected to the specified transmit and receive pins, with the specified baud.
+     *
+     *  @param tx Transmit pin
+     *  @param rx Receive pin
+     *  @param baud The baud rate of the serial port (optional, defaults to MBED_CONF_PLATFORM_DEFAULT_SERIAL_BAUD_RATE)
+     *
+     *  @note
+     *    Either tx or rx may be specified as NC if unused
+     */
+    UnbufferedSerial(PinName tx, PinName rx, int baud = MBED_CONF_PLATFORM_DEFAULT_SERIAL_BAUD_RATE);
+
+    virtual ssize_t write(const void *buffer, size_t size);
+    virtual ssize_t read(void *buffer, size_t size);
+    virtual off_t seek(off_t offset, int whence = SEEK_SET);
+    virtual off_t size();
+    virtual int isatty();
+    virtual int close();
+    virtual short poll(short events) const;
+
+#if !(DOXYGEN_ONLY)
+protected:
+
+    /* Acquire exclusive access to this serial port
+     */
+    virtual void lock(void);
+
+    /* Release exclusive access to this serial port
+     */
+    virtual void unlock(void);
+#endif
+};
+```
+
+# Usage scenarios and examples
+
+### Scenario 1 `Printing to stdout`
+
+There is no change in the current APi in regard to printing to stdout.  Applications should call `printf` directly as shown below. The console baud rate can be configured via `platform.stdio-baud-rate` in `mbed_app.json`.
+
+```C
+printf("Hello");
+```
+or
+```C
+fprintf(stdout, "Hello");
+```
+
+### Scenario 2 `Printing to a serial port other than stdout`
+
+The below example shows pseudocode for how to print `Hello` to `serial` port using the new API. `serial.printf` cannot be called directly; instead the application first calls _fdopen_ to get a _FILE*_ to use the printf function.
+
+```C
```

Comment:
Could we flip the order and give the most "straightforward" example first? That is the direct usage of the class and write. Also mentioning why would you choose to use the file way (I'm assuming redirection and extended formating). How does it affect the image size?

Reply:
Agreed.